# Multiple Linear regression 
# Çoklu Doğrusal REGREASYON
##🟨 2. Multiple Linear Regression (Çoklu Doğrusal Regresyon)
📌 Tanım:

Birden fazla bağımsız değişken ile bir bağımlı değişken arasındaki ilişkiyi modelliyorsun.
📈 Formül:

Y = b0 + b1*X1 + b2*X2 + ... + bn*Xn

🎯 Amaç:

Çoklu faktörleri hesaba katarak Y'yi tahmin etmek.
✅ Örnek:

    X1 = çalışma süresi (saat)

    X2 = önceki sınav ortalaması

    X3 = uyku süresi

    Y = sınav notu

    Amaç: Bu 3 değişkene göre notu tahmin etmek.

 | Model           | Bağımsız Değişken | İlişki Türü       | Kullanım Durumu                     |
| --------------- | ----------------- | ----------------- | ----------------------------------- |
| Simple Linear   | 1                 | Doğrusal (lineer) | X ile Y arasında basit ilişki varsa |
| Multiple Linear | 2+                | Doğrusal (lineer) | Çok faktör varsa                    |
| Polynomial      | 1 (ama X^2, X^3…) | Doğrusal değil    | Eğri (non-linear) bir ilişki varsa  |
        

## Kütüphaneleri Ekleme

In [10]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Veriseti Aktarımı Ve Temizleme

In [11]:
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

In [12]:
print(X)

[[165349.2 136897.8 471784.1 'New York']
 [162597.7 151377.59 443898.53 'California']
 [153441.51 101145.55 407934.54 'Florida']
 [144372.41 118671.85 383199.62 'New York']
 [142107.34 91391.77 366168.42 'Florida']
 [131876.9 99814.71 362861.36 'New York']
 [134615.46 147198.87 127716.82 'California']
 [130298.13 145530.06 323876.68 'Florida']
 [120542.52 148718.95 311613.29 'New York']
 [123334.88 108679.17 304981.62 'California']
 [101913.08 110594.11 229160.95 'Florida']
 [100671.96 91790.61 249744.55 'California']
 [93863.75 127320.38 249839.44 'Florida']
 [91992.39 135495.07 252664.93 'California']
 [119943.24 156547.42 256512.92 'Florida']
 [114523.61 122616.84 261776.23 'New York']
 [78013.11 121597.55 264346.06 'California']
 [94657.16 145077.58 282574.31 'New York']
 [91749.16 114175.79 294919.57 'Florida']
 [86419.7 153514.11 0.0 'New York']
 [76253.86 113867.3 298664.47 'California']
 [78389.47 153773.43 299737.29 'New York']
 [73994.56 122782.75 303319.26 'Florida']
 [67532

In [13]:
print(y)

[192261.83 191792.06 191050.39 182901.99 166187.94 156991.12 156122.51
 155752.6  152211.77 149759.96 146121.95 144259.4  141585.52 134307.35
 132602.65 129917.04 126992.93 125370.37 124266.9  122776.86 118474.03
 111313.02 110352.25 108733.99 108552.04 107404.34 105733.54 105008.31
 103282.38 101004.64  99937.59  97483.56  97427.84  96778.92  96712.8
  96479.51  90708.19  89949.14  81229.06  81005.76  78239.91  77798.83
  71498.49  69758.98  65200.33  64926.08  49490.75  42559.73  35673.41
  14681.4 ]


### State Sütununa Kodluyoruz
Encoding categorical data, burada 3. sütunu encodluyoruz

In [14]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [3])], remainder='passthrough')
X = np.array(ct.fit_transform(X))
print(X)

[[0.0 0.0 1.0 165349.2 136897.8 471784.1]
 [1.0 0.0 0.0 162597.7 151377.59 443898.53]
 [0.0 1.0 0.0 153441.51 101145.55 407934.54]
 [0.0 0.0 1.0 144372.41 118671.85 383199.62]
 [0.0 1.0 0.0 142107.34 91391.77 366168.42]
 [0.0 0.0 1.0 131876.9 99814.71 362861.36]
 [1.0 0.0 0.0 134615.46 147198.87 127716.82]
 [0.0 1.0 0.0 130298.13 145530.06 323876.68]
 [0.0 0.0 1.0 120542.52 148718.95 311613.29]
 [1.0 0.0 0.0 123334.88 108679.17 304981.62]
 [0.0 1.0 0.0 101913.08 110594.11 229160.95]
 [1.0 0.0 0.0 100671.96 91790.61 249744.55]
 [0.0 1.0 0.0 93863.75 127320.38 249839.44]
 [1.0 0.0 0.0 91992.39 135495.07 252664.93]
 [0.0 1.0 0.0 119943.24 156547.42 256512.92]
 [0.0 0.0 1.0 114523.61 122616.84 261776.23]
 [1.0 0.0 0.0 78013.11 121597.55 264346.06]
 [0.0 0.0 1.0 94657.16 145077.58 282574.31]
 [0.0 1.0 0.0 91749.16 114175.79 294919.57]
 [0.0 0.0 1.0 86419.7 153514.11 0.0]
 [1.0 0.0 0.0 76253.86 113867.3 298664.47]
 [0.0 0.0 1.0 78389.47 153773.43 299737.29]
 [0.0 1.0 0.0 73994.56 122782.75 3

## Verisetini train ve test olarak ayırıyoruz
Splitting the dataset into the training set and test set

In [15]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

## Çoklu doğrusal regrasyon modeli ile eğitim
kukla tuzayı ve model oluştururken öğrenmiş olduğumuz 5 maddeyi fonksiyonlar otomatik yapıyor sonuçta bu teknikleri uygulamak iş hayatında zaman kaybına sebeb olmakla birlikte amerikayı tekrar keşfetmeye gerek yoktur kütüphane sınıfları bunları hallediyor, bu durumda veri seti için en uygun modeli seçmek bize kalıyor

In [18]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train) #eğitiyoruz verisetini
#artık söz konusu modeli kullanarak yatırımcılar bu bilgilere dayanarak ne kadar 
#kar edeceklerini tahmin etmek için bu modeli girişimlerine uygulayabilirler

LinearRegression()

## Test setinden tahminler üretmek
Predicting the Test set results
Bu veristetinde birden fazla bağımsız değişken vardır özellikler x de

In [22]:
y_pred = regressor.predict(X_test)
np.set_printoptions(precision=2) #hassasiyet değeri 2
# tahmin edilen karları ve gerçek karları iki vektörünü yan yana göstereceğiz
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[103015.2  103282.38]
 [132582.28 144259.4 ]
 [132447.74 146121.95]
 [ 71976.1   77798.83]
 [178537.48 191050.39]
 [116161.24 105008.31]
 [ 67851.69  81229.06]
 [ 98791.73  97483.56]
 [113969.44 110352.25]
 [167921.07 166187.94]]


görüldüğü üzere iyi tahminler yapmış diyebiliriz

## Making a single prediction (for example the profit of a startup with R&D Spend = 160000, Administration Spend = 130000, Marketing Spend = 300000 and State = 'California')

In [24]:
print(regressor.predict([[1, 0, 0, 160000, 130000, 300000]]))

[181566.92]


Dolayısıyla modelimiz, Ar-Ge'ye 160.000, İdari İşlere 130.000 ve Pazarlamaya 300.000 dolar harcayan bir Kaliforniya girişiminin karının 181.566,92 dolar olacağını öngörüyor.

**Önemli not 1:** Özelliklerin değerlerinin hepsinin çift köşeli parantez içinde girildiğine dikkat edin. Bunun nedeni, "öngörme" yönteminin girdilerinin biçimi olarak her zaman 2 boyutlu bir dizi beklemesidir. Ve değerlerimizi çift köşeli parantez içine koymak girdiyi tam olarak 2 boyutlu bir dizi yapar. Basitçe söylemek gerekirse:

$1, 0, 0, 160000, 130000, 300000 \rightarrow \textrm{scalars}$

$[1, 0, 0, 160000, 130000, 300000] \rightarrow \textrm{1D array}$

$[[1, 0, 0, 160000, 130000, 300000]] \rightarrow \textrm{2D array}$

**Önemli not 2:** Ayrıca, "California" durumunun son sütunda bir dize olarak değil, ilk üç sütunda "1, 0, 0" olarak girildiğine dikkat edin. Bunun nedeni, tahmin yönteminin elbette durumun tek sıcak kodlanmış değerlerini beklemesidir ve X özellik matrisinin ikinci satırında gördüğümüz gibi, "California" "1, 0, 0" olarak kodlanmıştır. Ve bu değerleri son üç sütuna değil, ilk üç sütuna dahil etmeye dikkat edin, çünkü kukla değişkenler her zaman ilk sütunlarda oluşturulur.

## Getting the final linear regression equation with the values of the coefficients

In [26]:
print(regressor.coef_)
print(regressor.intercept_)

[ 8.66e+01 -8.73e+02  7.86e+02  7.73e-01  3.29e-02  3.66e-02]
42467.529248548584


Dolayısıyla çoklu doğrusal regresyon modelimizin denklemi şu şekildedir:

$$
Profit=86.6×Dummy State 1−873×Dummy State 2+786×Dummy State 3+0.773×R&D Spend+0.0329×Administration+0.0366×Marketing Spend+42467.53
$$

**Önemli Not:** Bu katsayıları elde etmek için regressor nesnemizden "coef_" ve "intercept_" niteliklerini çağırdık. Python'daki nitelikler yöntemlerden farklıdır ve genellikle basit bir değer veya bir dizi değer döndürür.